### ISSUE TITLE
Add link checker to avoid addition of duplicate links

### ISSUE THREAD

**Issue by sonylomo** [rupali-codes/LinksHub#408]
@sonylomo Sure ! It is indeed a credible idea. One way to implement this would be to have the system check the links that users enter against the existing links in the database. If a user tries to add a link that already exists in the database, the system can then display a message or warning that the resource has already been added. To implement this you could write a script that checks the link or related date to the entered data.

_Originally posted by @ujjawaltyagii in https://github.com/rupali-codes/LinksHub/issues/397#issuecomment-1500200000_

**Comment by Dikshantw** [comment]
we can add a URL checker to each category folder in database which will ask the user for which subcategory they want to add their links to and ask user for name, description, URL, category, subcategory, and language. If the URL already exists in that json file they will get a warning and if not the links and description will directly get added to the same json file that they wanted to add.
should I start working on it?

**Comment by Dikshantw** [comment]
@rupali-codes if we do this, we also need to make changes in contributing.md since all the appending of new link will be done by checker file.

**Comment by rupali-codes** [comment]
@Dikshantw Are you talking about making a form then append links from there?

**Comment by Dikshantw** [comment]
@rupali-codes No, but more like when user runs the checker file, in terminal they can add their links and the code will tell if it's already available or not and if not then gets added to the respective json file.
give me some time i will post some screenshots of the same.

**Comment by CBID2** [comment]
> _@sonylomo Sure ! It is indeed a credible idea. One way to implement this would be to have the system check the links that users enter against the existing links in the database. If a user tries to add a link that already exists in the database, the system can then display a message or warning that the resource has already been added. To implement this you could write a script that checks the link or related date to the entered data._
> 
> _Originally posted by @ujjawaltyagii in [#397 (comment)](https://github.com/rupali-codes/LinksHub/pull/397#issuecomment-1500200000)_

I like this idea! :)

**Comment by rupali-codes** [comment]
> @rupali-codes No, but more like when user runs the checker file, in terminal they can add their links and the code will tell if it's already available or not and if not then gets added to the respective json file. give me some time i will post some screenshots of the same.

Sure

**Comment by Dikshantw** [comment]
![if_exists](https://user-images.githubusercontent.com/122555738/231256901-6f30ae38-9695-4e57-b951-3e45b645e579.png)

**Comment by Dikshantw** [comment]
if the url exists:
![if_exists](https://user-images.githubusercontent.com/122555738/231257352-e9230814-dd05-437d-b0f8-504e86495c94.png)

if url doesn't exists:
![if_doesn't_exists](https://user-images.githubusercontent.com/122555738/231257554-59e3dd88-1ce9-4df9-9c02-3126b5b10ab9.png)

**Comment by Dikshantw** [comment]
if the url exists:
![if_exists](https://user-images.githubusercontent.com/122555738/231257352-e9230814-dd05-437d-b0f8-504e86495c94.png)

if url doesn't exists:
![if_doesn't_exists](https://user-images.githubusercontent.com/122555738/231257554-59e3dd88-1ce9-4df9-9c02-3126b5b10ab9.png)

**Comment by rupali-codes** [comment]
Nice, but it won't be sufficient if someone adds new Link using gitpod or github it self.

**Comment by Divyansh-0** [comment]
So can we try writing a script within data.json which maps over all values of link and make sure none is repeated?? Then we can use useState or something to display only incase theres no repeated links

**Comment by rupali-codes** [comment]
> So can we try writing a script within data.json which maps over all values of link and make sure none is repeated?? Then we can use useState or something to display only incase theres no repeated links

Sounds cool

**Comment by Divyansh-0** [comment]
Hey i have a better idea we can create a pre-commit hook which contains the url validation function. I think it would be the perfect solution incase people use GitPod or Github to add link . What do you guys think??

**Comment by rupali-codes** [comment]
> Hey i have a better idea we can create a pre-commit hook which contains the url validation function. I think it would be the perfect solution incase people use GitPod or Github to add link . What do you guys think??

Nice, then we will have to put a condition in there, but what if the link already exist? how we gonna protect?

**Comment by Divyansh-0** [comment]
> > Hey i have a better idea we can create a pre-commit hook which contains the url validation function. I think it would be the perfect solution incase people use GitPod or Github to add link . What do you guys think??
> 
> Nice, then we will have to put a condition in there, but what if the link already exist? how we gonna protect?

In that case we will have to write a script in pre-commit hook which looks for the link in all the json files to see if it already exists. and if link already exists then one cannt commit

**Comment by rupali-codes** [comment]
> > > Hey i have a better idea we can create a pre-commit hook which contains the url validation function. I think it would be the perfect solution incase people use GitPod or Github to add link . What do you guys think??
> > 
> > 
> > Nice, then we will have to put a condition in there, but what if the link already exist? how we gonna protect?
> 
> In that case we will have to write a script in pre-commit hook which looks for the link in all the json files to see if it already exists. and if link already exists then one cannt commit

Okay, you can make a PR and then we will see

**Comment by Isha988** [comment]
@rupali-codes @CBID2 I want to work on this issue under gssoc program.

**Comment by Isha988** [comment]
@CBID2 Thank you for assigning it to me. Is it ok if I start doing it after completing the voting functionality issue?

**Comment by CBID2** [comment]
> @CBID2 Thank you for assigning it to me. Is it ok if I start doing it after completing the voting functionality issue?

Finish this issue first @Isha988.

**Comment by Isha988** [comment]
> > @CBID2 Thank you for assigning it to me. Is it ok if I start doing it after completing the voting functionality issue?
> 
> Finish this issue first @Isha988.

Okay. I will do it first.

**Comment by Divyansh-0** [comment]
So are we going with the dataBase approach or you are writing some script to check duplicate links??

**Comment by Isha988** [comment]
> Hey i have a better idea we can create a pre-commit hook which contains the url validation function. I think it would be the perfect solution incase people use GitPod or Github to add link . What do you guys think??

@Divyansh-0 I am thinking of going with this approach, creating a pre commit hook in python language.

**Comment by Divyansh-0** [comment]
> > Hey i have a better idea we can create a pre-commit hook which contains the url validation function. I think it would be the perfect solution incase people use GitPod or Github to add link . What do you guys think??
> 
> @Divyansh-0 I am thinking of going with this approach, creating a pre commit hook in python language.

@Isha988  I tried that approach but it was a bit difficult to loop through all the json files to find a match . I had path conflict issues hope you can do it 🤞

**Comment by rupali-codes** [comment]
@Isha988 are you still working on the issue?

**Comment by Isha988** [comment]
> @Isha988 are you still working on the issue?

Yes, I am working but I am unable to make any solution till now.
